keep/action/action.py

```diff
@@ -81,13 +51,80 @@ def _check_throttling(self, action_name):
-            self.provider.notify(**rendered_value)
+        # the item holds the value we are going to iterate over
+        items = self.io_handler.render(self.config.get("foreach"))
+        # apply ALL conditions (the decision whether to run or not is made in the end)
+        for item in items:
+            self.context_manager.set_for_each_context(item)
+            self._run_single()
 
     def _run_single(self):
+        # First, execute all conditions
+
+        # Initialize all conditions
+        conditions = [
+            ConditionFactory.get_condition(
+                condition.get("type"),
+                condition.get("name", None) or condition.get("type"),
+                condition,
+            )
+            for condition in self.conditions
+        ]
+        for condition in conditions:
+            condition_compare_to = condition.get_compare_to()
+            condition_compare_value = condition.get_compare_value()
+            condition_result = condition.apply(
+                condition_compare_to, condition_compare_value
+            )
+            self.context_manager.set_condition_results(
+                self.name,
+                condition.condition_name,
+                condition.condition_type,
+                condition_compare_to,
+                condition_compare_value,
+                condition_result,
+                condition_alias=condition.condition_alias,
+                **condition.condition_context,
+            )
+
+        # Second, decide if need to run
+        # after all conditions are applied, check if we need to run
+        # there are 2 cases:
+        # 1. a "if" block is supplied, then use it
+        # 2. no "if" block is supplied, then use the AND between all conditions
+        if self.config.get("if"):
+            if_conf = self.config.get("if")
+        else:
+            # create a string of all conditions, separated by "&&"
+            if_conf = " and ".join(
+                [f"{{{{ {condition.condition_alias} }}}} " for condition in conditions]
+            )
+
+        # Now check it
+        if_met = self.io_handler.render(if_conf)
+        # if there's more than one condition, we need to evaluate the "if" string
+        # otherwise, it's already a boolean
+        if len(conditions) > 1:
+            if_met = eval(if_met)
+
+        if not if_met:
+            self.logger.info(
+                "Action %s evaluated NOT to run, Reason: %s evaluated to false.",
+                self.config.get("name"),
+                if_conf,
+            )
+            return
+
+        self.logger.info("Action %s evaluated to run!", self.config.get("name"))
+
+        # Third, check throttling
+        # Now check if throttling is enabled
+        throttled = self._check_throttling(self.config.get("name"))
+        if throttled:
+            self.logger.info("Action %s is throttled", self.config.get("name"))
+            return
```

Comment:
because "if" is "or"/"and" between conditions (e.g. "{{ A }} or {{ B}}") where each of A,B is evaluated to true/false.

Throttling is a logic (strategy) to decide if an action should be throttled or not (which is handled by `_check_throttling`)

Reply:
> Throttling is a logic (strategy) to decide if an action should be throttled or not (which is handled by _check_throttling)

which doesn't return true/false? not sure I fully understand but if it makes sense then 👌🏼

Reply:
it does, but its independent of the other "if" block. the original "if" is used so you'll be able to do "OR" between conditions. the throttling is binary - whether true or false

Reply:
still don't understand but your call, resolving this since i don't have any other input